addons/udes_advanced_packaging/tests/test_package_reservation.py

```diff
@@ -0,0 +1,172 @@
+            lot_name="sn01",
+        )
+        strawberry_quant_2 = self.create_quant(
+            self.strawberry.id,
+            self.test_stock_location_01.id,
+            1,
+            package_id=package.id,
+            lot_name="sn02",
+        )
+        all_quants = apple_quant + strawberry_quant_1 + strawberry_quant_2
+
+        # all quants should be unreserved
+        self.assertTrue(
+            all([q.reserved_quantity == 0 for q in all_quants]), "One or more quants are reserved"
+        )
+
+        # create a Pick for 1 apple
+        create_info = [{"product": self.apple, "uom_qty": 1}]
+        picking = self.create_picking(
+            self.picking_type_pick, products_info=create_info, confirm=True, assign=False
+        )
+        picking.action_assign()
+
+        # apple quant should be fully reserved
+        self.assertEqual(apple_quant.reserved_quantity, 3)
+        # strawberry quants should be fully reserved
+        self.assertEqual(strawberry_quant_1.reserved_quantity, 1)
+        self.assertEqual(strawberry_quant_2.reserved_quantity, 1)
+
+        # apple stock.move.lines of the picking should have the package
+        apple_mls = picking.mapped("move_line_ids").filtered(lambda ml: ml.product_id == self.apple)
+        self.assertTrue(all([ml.package_id == package for ml in apple_mls]))
+
+        # strawberry stock.move.lines of the picking should have the package
+        strawberry_mls = picking.mapped("move_line_ids").filtered(
+            lambda ml: ml.product_id == self.strawberry
+        )
+        self.assertTrue(all([ml.package_id == package for ml in apple_mls]))
+
+    def test_partial_reservation_package_one_still_functional(self):
+        """Test to reserve a full package for one product
+        when initial demand < package quantity
+        """
+
+        # disable full package reservation
+        self.picking_type_pick.u_reserve_as_packages = False
+
+        # create a package
+        package = self.create_package()
+
+        # create a quant of 4 apples in a stock sublocation for the package
+        quant = self.create_quant(
+            self.apple.id, self.test_stock_location_01.id, 4, package_id=package.id
+        )
+        # the quant should be unreserved
+        self.assertEqual(quant.reserved_quantity, 0, "The quant is reserved")
+
+        # create a Pick for 1 apple
+        create_info = [{"product": self.apple, "uom_qty": 1}]
+        picking = self.create_picking(
+            self.picking_type_pick, products_info=create_info, confirm=True, assign=False
+        )
+        picking.action_assign()
+
+        # the quant should be partially reserved
+        self.assertEqual(quant.reserved_quantity, 1)
+
+        # apple stock.move.lines of the picking should have the package
+        apple_mls = picking.mapped("move_line_ids").filtered(lambda ml: ml.product_id == self.apple)
+        self.assertTrue(all([ml.package_id == package for ml in apple_mls]))
```

Comment:
nit: Expect one move line so don't need the list comprehension assertTrue behaviour, assertEqual would be clearer

Reply:
We can have more than one move line if we have more than one move because they are not merged on action_confirm, once that is resolved we should have one line only.

Reply:
Merging is fixed by this line in `_prepare_move` in `udes_stock`: https://github.com/unipartdigital/udes-open/pull/554/files#diff-bd6ad1e2e7579def4095cbf373ff632d73cf4bcb1c1c3cc55f8299eaaf72accfR368